src/core/router/transports/ssu/session.cc

```diff
@@ -488,88 +488,146 @@ void SSUSession::ProcessSessionCreated(const SSUPacket* packet)
+
+      // Set Alice's IP address size and address
+      auto const alice_ip(
+          core::AddressToByteVector(m_RemoteEndpoint.address()));
+
+      message.SetIPAddress(
+          // TODO(unassigned): remove const_cast, see bytestream TODO
+          const_cast<std::uint8_t*>(alice_ip.data()),
+          alice_ip.size());  // message IP address size must be set internally
+
+      // Set Alice's port
+      message.SetPort(m_RemoteEndpoint.port());
+
+      // Compute exchanged session dataset size
+      // TODO(anonimal): helper function to calculate size
+      // TODO(anonimal): at this point, why would we allow mix-and-match IPv6 to send to IPv4 - or vice versa...
+      bool const is_IPv6 = address->host.is_v6();
+      std::uint16_t const data_size = DHKeySize::PubKey * 2  // DH X+Y
+                                      + alice_ip.size()  // Alice's address
+                                      + 2  // Alice's port
+                                      + (is_IPv6 ? 16 : 4)  // Bob's address
+                                      + 2  // Bob's port
+                                      + 4  // Alice's relay tag
+                                      + 4;  // Bob's signed-on time
+
+      // Prepare dataset of exchanged session data (the dataset we will sign)
+      // TODO(anonimal): assert for bad design. Redesign.
+      assert(!m_SessionConfirmData.size());
+      m_SessionConfirmData.reserve(data_size);
+
+      core::OutputByteStream data(m_SessionConfirmData.data(), data_size);
+
+      // Alice's DH X
+      data.WriteData(dh_x, SSUSize::DHPublic);
+
+      // Our (Bob's) DH Y
+      data.WriteData(message.GetDhY(), SSUSize::DHPublic);
+
+      // Alice's address and port
+      data.WriteData(alice_ip.data(), alice_ip.size());
+      data.Write<std::uint16_t>(message.GetPort());
+
+      // Our (Bob's) address
+      data.WriteData(
+          is_IPv6 ? address->host.to_v6().to_bytes().data()
+                  : address->host.to_v4().to_bytes().data(),
+          is_IPv6 ? 16 : 4);
+
+      // Our (Bob's) port
+      data.Write<std::uint16_t>(address->port);
+
+      // Set Alice's relay tag
+      std::uint32_t relay_tag = 0;
+      if (context.GetRouterInfo().HasCap(RouterInfo::Cap::SSUIntroducer))
+        {
+          // Non-zero = we are offering ourselves to be an introducer
+          relay_tag = core::Rand<std::uint32_t>();
+          if (!relay_tag)
+            {
+              // TODO(anonimal): ...not good if should we have more than one relay
+              //  with tag valued 1. Get existing tags and set appropriately.
+              relay_tag = 1;
+            }
+          m_Server.AddRelay(relay_tag, m_RemoteEndpoint);
+        }
+      message.SetRelayTag(relay_tag);
+      data.Write<std::uint32_t>(relay_tag);
+
+      // Our (Bob's) signed-on time
+      message.SetSignedOnTime(core::GetSecondsSinceEpoch());
+      data.Write<std::uint32_t>(message.GetSignedOnTime());
```

Comment:
Do we need to be concerned about thread safety here? Is there any danger of `GetSignedOnTime` returning/being called before `SetSignedOnTime` returns?

Reply:
...

No. This object is created locally, within the function, within the thread. If the 2nd line was executed before the 1st, threading would be the last of our concerns since the compiler would be broken beyond recognition.

Look at the transports impl for answers to future threading questions.